src/login/loginCoordinator.ts

```diff
@@ -191,54 +201,93 @@ export class LoginCoordinator {
 	private async attemptLogin(
 		deployment: Deployment,
 		isAutoLogin: boolean,
+		providedToken?: string,
 	): Promise<LoginResult> {
-		const needsToken = needToken(vscode.workspace.getConfiguration());
 		const client = CoderApi.create(deployment.url, "", this.logger);
 
-		let storedToken: string | undefined;
-		if (needsToken) {
-			const auth = await this.secretsManager.getSessionAuth(
-				deployment.safeHostname,
+		// mTLS authentication (no token needed)
+		if (!needToken(vscode.workspace.getConfiguration())) {
+			return this.tryMtlsAuth(client, isAutoLogin);
+		}
+
+		// Try provided token first
+		if (providedToken) {
+			const result = await this.tryTokenAuth(
+				client,
+				providedToken,
+				isAutoLogin,
 			);
-			storedToken = auth?.token;
-			if (storedToken) {
-				client.setSessionToken(storedToken);
+			if (result !== "retry") {
```

Comment:
To me "retry" sounds more like "try this token again later" but if I understand correctly it means to try the next token.  Could we return "next" or something like that?  Or add the status code or `reason` to a failed `LoginResult` and do `code !== 401` or something.